File: C++/15.命令模式/CommandPatternDemo.cpp
/**
 * @file CommandPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-14
 *
 * @copyright Copyright (c) 2019
 *
 */
#include "Broker.h"
#include "BuyStock.h"
#include "SellStock.h"

int main(int argc, char const* argv[]) {
    std::shared_ptr<Stock> abc_stock = std::make_shared<Stock>();
    std::shared_ptr<BuyStock> buy_stock_order = std::make_shared<BuyStock>(abc_stock);
    std::shared_ptr<SellStock> sell_stock_order = std::make_shared<SellStock>(abc_stock);

    std::shared_ptr<Broker> broker = std::make_shared<Broker>();
    broker->TakeOrder(buy_stock_order);
    broker->TakeOrder(sell_stock_order);
    broker->PlaceOrder();

    return 0;
}

File: Objective-C/15.CommandPattern/CommandPatternDemo.m
//
//  CommandPatternDemo.m
//  15.CommandPattern
//
//  Created by Layne on 2020/11/5.
//

#import "Broker.h"
#import "BuyStock.h"
#import "SellStock.h"
#import "Stock.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        Stock* abcStock = [[Stock alloc] init];

        BuyStock* buyStock = [[BuyStock alloc] initWithStock:abcStock];
        SellStock* sellSock = [[SellStock alloc] initWithStock:abcStock];

        Broker* broker = [[Broker alloc] init];
        [broker takeOrder:buyStock];
        [broker takeOrder:sellSock];

        [broker placeOrders];
    }
    return 0;
}
